myia/clone.py

```diff
@@ -0,0 +1,227 @@
+    added with the `add_clone` method, which offers more options.
+
+    Cloning a graph automatically involves cloning every graph which
+    is nested in it.
+
+    To get the clone of a graph, index the GraphCloner with the graph,
+    e.g. `cloned_g = graph_cloner[g]`
+
+    Attributes:
+        total: Whether to clone every graph encountered when walking
+            through the graphs to clone, even if these graphs are
+            not nested.
+        clone_constants: Whether to clone Constant nodes.
+        relation: The relation the cloned nodes present with respect
+            to the originals, for debugging purposes. Default is
+            'copy'.
+
+    """
+
+    def __init__(self,
+                 *graphs: Graph,
+                 total: bool = False,
+                 clone_constants: bool = False,
+                 relation: str = 'copy') -> None:
+        """Initialize a GraphCloner."""
+        self.todo: Set[Graph] = set()
+        self.graph_mappings: Dict[Graph, Union[Graph, bool]] = {}
+        self.repl: Dict[Union[Graph, ANFNode],
+                        Union[Graph, ANFNode]] = {None: None}
+        self.total = total
+        self.clone_constants = clone_constants
+        self.relation = relation
+        for graph in graphs:
+            self.add_clone(graph)
+
+    def _add_clone(self,
+                   graph: Graph,
+                   target_graph: Graph = None,
+                   set_output: bool = None) -> Graph:
+        """Register a clone for the given graph.
+
+        This does not register graphs nested in the given graph, so
+        use this method with caution.
+
+        Arguments:
+            graph: The graph to clone.
+            target_graph: The graph which will contain the clones of
+                the nodes in the original graph. If target_graph is
+                None or not given, a new Graph will be created to
+                serve as the target_graph.
+            set_output: Whether the output of the target_graph should
+                be set to the clone of the output of the original
+                graph. If set_output is None or not given, then it
+                is true if target_graph is None (and thus must be
+                created), false otherwise.
+
+        Return:
+            target_graph, or the clone associated to the original graph.
+        """
+        if set_output is None:
+            set_output = target_graph is None
+
+        if target_graph is None:
+            with About(graph.debug, self.relation):
+                target_graph = Graph()
+
+        self.repl[graph] = target_graph
+        if set_output:
+            self.graph_mappings[graph] = target_graph
+        else:
+            self.graph_mappings[graph] = True
+        self.todo.add(graph)
+
+        return target_graph
+
+    def add_clone(self,
+                  graph: Graph,
+                  target_graph: Graph = None,
```

Comment:
I don't see any point where this parameter has a non-default value.  I question its usefulness.

Reply:
It will have a non-default value whenever this is used for inlining, which is actually most of the time. If you wanted to inline an application of `f` in `g`, you would call `add_clone(f, g, apply.inputs)`. I just didn't add the inlining optimizations in yet.